sway-lib-std/src/vm/evm/ecr.sw

```diff
@@ -0,0 +1,32 @@
+library ecr;
+
+use ::address::Address;
+use ::b512::B512;
+use ::context::registers::error;
+use ::ecr::{EcRecoverError, ec_recover};
+use ::hash::{HashMethod, hash_pair};
+use ::result::*;
+
+/// Recover the address derived from the private key used to sign a message.
+/// Returns a `Result` to let the caller choose an error handling strategy.
+/// Ethereum addresses are 20 bytes long, so these are left-padded to fit in a 32 byte Address type.
+pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Result<Address, EcRecoverError> {
```

Comment:
cc @nfurfaro @Braqzen for thoughts here. Do we want a new type like `EVMAddress`, rather than re-using `Address`?

We could just keep it as-is and change it later if we find out by using it that it's not the best.

Reply:
What is the additional benefit of implementing and maintaining a new type other than `Address`?
We already have `Address` and `ContractID`.
Breaking down `Address` into 2 types for namespacing alone is not something that I am fond of so if `Address` has all the functionality needed then I'd rather not have an additional type with a different name.

If there is some functionality / distinction which I am unaware of then I have no issue with an additional type for its explicit purpose.
FYI: I'm aware of ecrecover but I've never used it myself so @nfurfaro holds more weight here.

Reply:
I think unless we implement a B160 type, making the distinction doesn't give us much. Arguably the same could be said of zeroing out the first 12 bytes (we could just do the keccak hash of the pubkey and be done with). But I suppose it does make it more obvious visually what you're dealing with if there are 24 leading zeros.

Reply:
I don't think another new type is warranted here... That said, I can imagine a scenario where someone sends funds to an Ethereum address mistakenly because it's an `Address`.
I'm in favour of the zeroed initial bytes for the visual distinction they add. One could also implement a helper function in the Ethereum lib like `is_eth_address() ` which checks the initial 12 bytes.

Reply:
> `is_eth_address()`

If we go down this route of having to distinguish between a Fuel address and an Ethereum address then we should move that to the type system.

Reply:
I guess functionally there's no difference between sending to an Ethereum address by mistake and sending to the wrong fuel address. Both are on the user and should be avoided.

Reply:
I'm in favour of different types long term. It's worth bikeshedding over at some point so we may want to create an issue for it.

Semantically, Address and EthAddress are separate things with different "address spaces", just like ContractId and Address, though they share many properties (a trait maybe?)

Reply:
In favor of different types for safety. An Ethereum address is going to be potentially managed and used differently.

If there is a case where you want to cast an Ethereum address to an address there should be an explicit step and method for this.